src/platform/macos/events_loop/runloop_context.rs

```diff
@@ -0,0 +1,276 @@
+        let mut inner: Option<InnerRunloop> = Some(InnerRunloop::new(shared));
+
+        // Create a stack for it
+        let stack = context::stack::ProtectedFixedSizeStack::new(STACK_SIZE)
+            .expect("Runloop coroutine stack allocation");
+
+        // Set up a new context
+        let result = unsafe {
+            // Start by calling inner_runloop_entrypoint
+            let ctx = context::Context::new(&stack, inner_runloop_entrypoint);
+
+            // Yield to the coroutine, giving it a pointer to the inner runloop, and wait for it come back
+            ctx.resume(&mut inner as *mut Option<InnerRunloop> as usize)
+        };
+
+        Runloop{
+            _stack: stack,
+            ctx: Some(result.context),
+            _timer: Timer::new(0.005),
+            _observer: RunloopObserver::new(),
+        }
+    }
+
+    // Work the runloop, attempting to respect the timeout
+    pub fn work(&mut self, timeout: Timeout) {
+        unsafe {
+            if !msg_send![cocoa::base::class("NSThread"), isMainThread] {
+                panic!("Events can only be polled from the main thread on macOS");
+            }
+        }
+
+        // Make an Option<Timeout> that contains the timeout
+        // The coroutine will .take() it as soon as it returns (see InnerRunloop::yield_to_caller())
+        let mut timeout: Option<Timeout> = Some(timeout);
+
+        // Resume the coroutine, giving it a pointer to our local timeout
+        let context = self.ctx.take().expect("coroutine context");
+        let result = unsafe {
+            context.resume(&mut timeout as *mut Option<_> as usize)
+        };
+
+        // Store the new coroutine context
+        self.ctx = Some(result.context);
+
+        // Return to caller
+    }
+
+    // Attempt to wake the Runloop. Must be thread safe.
+    pub fn wake() {
+        // Try to context switch back to the main thread
+        if yield_to_caller() {
+            // We did!
+        } else {
+            // Wake the runloop, because... we can, I guess?
+            unsafe {
+                core_foundation::runloop::CFRunLoopWakeUp(core_foundation::runloop::CFRunLoopGetMain());
+            }
+        }
+    }
+}
+
+thread_local!{
+    // If we are inside the inner runloop, this contains the caller's context and their Timeout
+    static INSIDE_INNER_RUNLOOP_CONTEXT: Cell<Option<context::Context>> = Cell::new(None);
+    static INSIDE_INNER_RUNLOOP_TIMEOUT: Cell<Option<Timeout>> = Cell::new(None);
+}
+
+// This is the first function called from inside the coroutine. It must not return.
+// Contract: t.data is a *mut Option<InnerRunloop>.
+extern fn inner_runloop_entrypoint(t: context::Transfer) -> ! {
+    let inner: *mut Option<InnerRunloop> = t.data as _;
+
+    // Turn this into a mutable borrow, then move the inner runloop into the coroutine's stack
+    let mut inner: InnerRunloop =
+        unsafe { mem::transmute::<*mut Option<_>, &mut Option<_>>(inner) }
```

Comment:
I chose to see this as a chain of `&mut` -> `*mut` -> `usize` -> `*mut` -> `&mut` casts. The thread always goes straight down that line, executing everything top to bottom, and I opted to reinforce that idea by fearlessly reaching for `transmute`.

`transmute` is a bold statement, but so is "these things are guaranteed to happen in this order, so I know for a fact that this `usize` is actually a `&mut Option`". My opinion was that this particular construction improves clarity here, especially when the alternative is `as_mut() -> Option<&mut Option<T>>` in a confusing context where we've already got `&mut`s and `Option`s in play.

Having said all that, if you want me to change it, let me know :-)